lib/main.js

```diff
@@ -121,6 +121,12 @@ function initSession(shiny, options, shinyServer) {
 
       factory(url, ctx, (err, conn) => {
         pc.resolve(err, conn);
+        conn.onclose = e => {
```

Comment:
Aron's right; since `conn` is the final connection that's ultimately handed back, it's not safe to set `onclose` on it. Either you're overwriting someone else, or they're overwriting you. In this case, Shiny's own `onclose` is overwritten by yours, and that means that 1) Shiny's own reconnect mechanism doesn't work, and 2) we're supposed to trigger a jquery event `"shiny:disconnected"` on the document and that gets squelched.

You can see this yourself by modifying a Shiny app to contain the line `session$allowReconnect(TRUE)` just inside the server function `function(input, output, session) {...}`. Then load up the app, and then kill the server. Once the shiny-server-client reconnect gives up (you can accelerate this by restarting the server, it'll quickly realize its old session was gone) you'll see Shiny attempt to start a new session with the current input values.

Probably the best thing to do is as Aron said, and create another decorator that does nothing but listen on `onclose` and call `this._ctx.emit("disconnect", e)`.

Sorry that the two reconnect methods haven't been properly documented; the user-facing documentation is coming from @wch any day now and will be on http://shiny.rstudio.com/articles.